### PR TITLE
[OptionalOptimization]: LayerNorm forward Optimization with Welford

### PR DESCRIPTION
**Pull request by JamesLim-sy**
### PR types

Function optimization
### PR changes

OPs
### Describe

- Feature: Optimizing the layer_norm forward implement by welford method, and vectorization I/O. Duing to diffeten reduction operation of each colume from the original one, namely `LayerNormForward`  CUDA kernel, **then a `gflag` is set for optional usage**.

- Accuracy & Performance : Typical layer_norm forward case in AF2 model is below.

x_shape | norm_axis | dtype | PaddleDev/ms     Speed | PaddleOpt/ms     Speed | Diff | PaddleDev     Accuracy | PaddleOpt     Accuracy
-- | -- | -- | -- | -- | -- | -- | --
[65536, 64] | 1 | fp32 | 79.25 | 46.72 | 41.04% | 1.0E-07 | 1.0E-07
  |   | fp16 | 68.54 | 43.69 | 36.25% | 1.0E-04 | 1.0E-04
[65536, 128] | 1 | fp32 | 79.62 | 57.26 | 28.08% | 1.0E-07 | 1.0E-07
  |   | fp16 | 77.97 | 51.72 | 33.67% | 1.0E-04 | 1.0E-04
[32768, 256] | 1 | fp32 | 58.09 | 56.06 | 3.49% | 1.0E-07 | 1.0E-07
  |   | fp16 | 46.04 | 41.45 | 9.99% | 1.0E-04 | 1.0E-04
[262144, 64] | 1 | fp32 | 294.99 | 167.97 | 43.06% | 1.0E-07 | 1.0E-07
  |   | fp16 | 287.17 | 165.50 | 42.37% | 1.0E-04 | 1.0E-04

### CODE REVIEW

File: paddle/phi/kernels/gpu/layer_norm_kernel.cu [hunk 1]
```diff
@@ -13,14 +13,372 @@
...
+template <typename T, typename U, int VecSize>
+struct LayerNormDataReader {
+  __device__ __forceinline__ void operator()(const T *__restrict__ row_src,
+                                             U *buffer,
+                                             const int read_times,
+                                             const int cols_this_thread) {
+    using VecT = phi::AlignedVector<T, VecSize>;
+    const VecT *__restrict__ v_src =
+        reinterpret_cast<const VecT *__restrict__>(row_src);
+#pragma unroll
```
Comment (shaojiewang) [inline paddle/phi/kernels/gpu/layer_norm_kernel.cu]: If ```read_times``` is a run-time defined variable, unroll hint could be removed.

File: paddle/phi/kernels/gpu/layer_norm_kernel.cu [hunk 2]
```diff
@@ -13,14 +13,372 @@
...
+    for (int i = 0; i < read_times; ++i) {
+      VecT temp_src = v_src[threadIdx.x * read_times + i];
```
Comment (shaojiewang) [inline paddle/phi/kernels/gpu/layer_norm_kernel.cu]: [ discussion ] If ```read_times``` were larger than 1, loading may not always be coalesced.

File: paddle/phi/kernels/gpu/layer_norm_kernel.cu [hunk 3]
```diff
@@ -13,14 +13,372 @@
...
+                           U *mean_data,
+                           U *var_data,
+                           float epsilon,
+                           const int64_t rows,
+                           const int cols,
+                           const bool valid_scale,
+                           const bool valid_bias,
+                           const bool is_same_type) {
+  constexpr int WarpSize = 32;
+  constexpr int RowPerBlock = 4;
```
Comment (shaojiewang) [inline paddle/phi/kernels/gpu/layer_norm_kernel.cu:375]: ```RowsPerBlock``` perhaps :)

**Review by shaojiewang — COMMENTED** [review]
Code looks good. Just left some discussions.

File: paddle/phi/kernels/gpu/layer_norm_kernel.cu (hunk 1 above)
Comment (JamesLim-sy) [reply paddle/phi/kernels/gpu/layer_norm_kernel.cu]: I just add this `unroll`  without much consideration. Will be removed later.

File: paddle/phi/kernels/gpu/layer_norm_kernel.cu (hunk 2 above)
Comment (JamesLim-sy) [reply paddle/phi/kernels/gpu/layer_norm_kernel.cu]: yes, coalesced loading has not been accomplished yet, will be fixed later.

File: paddle/phi/kernels/gpu/layer_norm_kernel.cu [hunk 4]
```diff
@@ -13,14 +13,445 @@
...
 
 #include "paddle/phi/kernels/layer_norm_kernel.h"
-
+#include "gflags/gflags.h"
 #include "paddle/phi/backends/gpu/gpu_context.h"
 #include "paddle/phi/core/kernel_registry.h"
 #include "paddle/phi/kernels/funcs/layer_norm_impl.cu.h"
 #include "paddle/phi/kernels/funcs/layer_norm_util.h"
 
+DECLARE_bool(use_fast_math);
```
Comment (AnnaTrainingG) [inline paddle/phi/kernels/gpu/layer_norm_kernel.cu:22]: 感觉flag有些不明确，还以为是针对所有算子

Comment (JamesLim-sy) [reply paddle/phi/kernels/gpu/layer_norm_kernel.cu:22]: 在绝大部分case中，layerNormWithWelford算法的精度都能够满足要求，但仍由于这种计算中的reduction模式与其他计算不同，因此在暂时仍采用保守的模式，配备选择性优化开关，当精度测试满足模型要求之后，遂开启。

File: paddle/phi/kernels/gpu/layer_norm_kernel.cu [hunk 5]
```diff
@@ -13,14 +13,445 @@
...
+    WelfordOnline<U>(b_mean, b_square, b_cnt, mean, square, count);
+  }
+
+  *mean = __shfl_sync(0xffffffff, *mean, 0, kWarpSize);
+  *square = __shfl_sync(0xffffffff, *square, 0, kWarpSize);
+  *count = __shfl_sync(0xffffffff, *count, 0, kWarpSize);
+}
+
+template <int VecSize>
+struct ThreadAssigner {
```
Comment (ZzSean) [inline paddle/phi/kernels/gpu/layer_norm_kernel.cu:68]: 这个默认的功能和下面VecSize=1的功能看起来好像不太一样？

Comment (JamesLim-sy) [reply paddle/phi/kernels/gpu/layer_norm_kernel.cu:68]: 嗯，VecSize = 1 的情况下，warp内并不是全部线程都工作，比如 cols = 133时，则cols_per_thread = 5，而仅仅有前27个线程参与工作，其tid 为 0~25的前26个线程，每个线程处理5个数据；tid为26的线程，用于处理最后的三个数据。面对这种情况就需要判断每个线程需要处理多少个数据了，即计算 cols_this_thread.
对于VecSize > 1的情况，warp内每个线程都是均匀的处理相同数量的数据，此时cols_per_thread 也代表了 cols_this_thread

**Review by shaojiewang — APPROVED** [review]
LGTM

**Review by ZzSean — APPROVED** [review]
LGTM

**Pull request merged**